File: python/deflatedecompress.py
# 
# Simple DEFLATE decompressor (Python)
# 
# Copyright (c) Project Nayuki
# MIT License. See readme file.
# https://www.nayuki.io/page/simple-deflate-decompressor
# 

import io
from typing import BinaryIO, Dict, List, Optional, Sequence, Tuple


class BitInputStream:
	
	"""A stream of bits that can be read. Because they come from an underlying byte stream,
	the total number of bits is always a multiple of 8. Bits are packed in little endian within
	a byte. For example, the byte 0x87 reads as the sequence of bits [1,1,1,0,0,0,0,1]."""
	
	
	# ---- Fields ----
	
	# The underlying byte stream to read from.
	_input: BinaryIO
	
	# Either in the range [0x00, 0xFF] if bits are available, or -1 if end of stream is reached.
	_current_byte: int
	
	# Number of remaining bits in the current byte, in the range [0, 8).
	_num_bits_remaining: int
	
	
	# ---- Constructor ----
	
	def __init__(self, inp: BinaryIO):
		"""Constructs a bit input stream based on the given byte input stream."""
		self._input = inp
		self._current_byte = 0
		self._num_bits_remaining = 0
	
	
	# ---- Methods ----
	
	def get_bit_position(self) -> int:
		"""Returns the current bit position, which ascends from 0 to 7 as bits are read."""
		assert 0 <= self._num_bits_remaining <= 7, "Unreachable state"
		return -self._num_bits_remaining % 8
	
	
	def read_bit_maybe(self) -> int:
		"""Reads a bit from this stream. Returns 0 or 1 if a bit is available, or -1 if
		the end of stream is reached. The end of stream always occurs on a byte boundary."""
		if self._current_byte == -1:
			return -1
		if self._num_bits_remaining == 0:
			b: bytes = self._input.read(1)
			if len(b) == 0:
				self._current_byte = -1
				return -1
			self._current_byte = b[0]
			self._num_bits_remaining = 8
		assert self._num_bits_remaining > 0, "Unreachable state"
		self._num_bits_remaining -= 1
		return (self._current_byte >> (7 - self._num_bits_remaining)) & 1
	
	
	def read_uint(self, numbits: int) -> int:
		"""Reads the given number of bits from this stream,
		packing them in little endian as an unsigned integer."""
		if numbits < 0:
			raise ValueError("Number of bits out of range")
		result: int = 0
		for i in range(numbits):
			bit: int = self.read_bit_maybe()
			if bit == -1:
				raise EOFError("Unexpected end of stream")
			result |= bit << i
		return result
	
	
	def close(self) -> None:
		"""Closes this stream and the underlying input stream."""
		self._input.close()
		self._current_byte = -1
		self._num_bits_remaining = 0



class CanonicalCode:
	
	"""A canonical Huffman code, where the code values for each symbol is
	derived from a given sequence of code lengths. This data structure is
	immutable. This could be transformed into an explicit Huffman code tree.
	
	Example:
	  Code lengths (canonical code):
	    Symbol A: 1
	    Symbol B: 0 (no code)
	    Symbol C: 3
	    Symbol D: 2
	    Symbol E: 3
	  
	  Generated Huffman codes:
	    Symbol A: 0
	    Symbol B: (Absent)
	    Symbol C: 110
	    Symbol D: 10
	    Symbol E: 111
	  
	  Huffman code tree:
	      .
	     / \
	    A   .
	       / \
	      D   .
	         / \
	        C   E"""
	
	
	# ---- Field ----
	
	# This dictionary maps Huffman codes to symbol values. Each key is the
	# Huffman code padded with a 1 bit at the beginning to disambiguate codes
	# of different lengths (e.g. otherwise we can't distinguish b01 from
	# 0b0001). For the example of codelengths=[1,0,3,2,3], we would have:
	#     0b1_0 -> 0
	#    0b1_10 -> 3
	#   0b1_110 -> 2
	#   0b1_111 -> 4
	_code_bits_to_symbol: Dict[int,int]
	
	
	# ---- Constructor ----
	
	def __init__(self, codelengths: Sequence[int]):
		"""Constructs a canonical Huffman code from the given list of symbol code lengths.
		Each code length must be non-negative. Code length 0 means no code for the symbol.
		The collection of code lengths must represent a proper full Huffman code tree.
		Examples of code lengths that result in correct full Huffman code trees:
		- [1, 1] (result: A=0, B=1)
		- [2, 2, 1, 0, 0, 0] (result: A=10, B=11, C=0)
		- [3, 3, 3, 3, 3, 3, 3, 3] (result: A=000, B=001, C=010, ..., H=111)
		Examples of code lengths that result in under-full Huffman code trees:
		- [0, 2, 0] (result: B=00, unused=01, unused=1)
		- [0, 1, 0, 2] (result: B=0, D=10, unused=11)
		Examples of code lengths that result in over-full Huffman code trees:
		- [1, 1, 1] (result: A=0, B=1, C=overflow)
		- [1, 1, 2, 2, 3, 3, 3, 3] (result: A=0, B=1, C=overflow, ...)"""
		
		# Check argument values
		if any(x < 0 for x in codelengths):
			raise ValueError("Negative code length")
		
		self._code_bits_to_symbol = {}
		
		# Allocate code values to symbols. Symbols are processed in the order
		# of shortest code length first, breaking ties by lowest symbol value.
		nextcode: int = 0
		for codelength in range(1, max(codelengths) + 1):
			nextcode <<= 1
			startbit: int = 1 << codelength
			for (symbol, cl) in enumerate(codelengths):
				if cl != codelength:
					continue
				if nextcode >= startbit:
					raise ValueError("This canonical code produces an over-full Huffman code tree")
				self._code_bits_to_symbol[startbit | nextcode] = symbol
				nextcode += 1
		if nextcode != 1 << max(codelengths):
			raise ValueError("This canonical code produces an under-full Huffman code tree")
	
	
	# ---- Methods ----
	
	def decode_next_symbol(self, inp: BitInputStream) -> int:
		"""Decodes the next symbol from the given bit input stream based on this
		canonical code. The returned symbol value is in the range [0, len(codelengths))."""
		codebits: int = 1  # The start bit
		while True:
			# Accumulate one bit at a time on the right side until a match is found
			# in the code_bits_to_symbol dictionary. Because the Huffman code tree is
			# full, this loop must terminate after at most max(codelengths) iterations.
			codebits = codebits << 1 | inp.read_uint(1)
			result: int = self._code_bits_to_symbol.get(codebits, -1)
			if result != -1:
				return result
	
	
	def __str__(self) -> str:
		"""Returns a string representation of this canonical code,
		useful for debugging only, and the format is subject to change."""
		return "\n".join(
			"Code {}: Symbol {}".format(bin(codebits)[3 : ], symbol)
			for (codebits, symbol) in sorted(self._code_bits_to_symbol.items())) + "\n"



class ByteHistory:
	
	"""Stores a finite recent history of a byte stream. Useful as an implicit
	dictionary for Lempel-Ziv schemes. Mutable and not thread-safe."""
	
	
	# ---- Fields ----
	
	# Circular buffer of byte data.
	_data: List[int]
	
	# Index of next byte to write to, always in the range [0, len(_data)).
	_index: int
	
	# Number of bytes written, saturating at len(_data).
	_length: int
	
	
	# ---- Constructor ----
	
	def __init__(self, size: int):
		"""Constructs a byte history of the given size."""
		if size < 1:
			raise ValueError("Size must be positive")
		self._data = [0] * size
		self._index = 0
		self._length = 0
	
	
	# ---- Methods ----
	
	def append(self, b: int) -> None:
		"""Appends the given byte to this history.
		This overwrites the byte value at `size` positions ago."""
		assert 0 <= self._index < len(self._data), "Unreachable state"
		self._data[self._index] = b
		self._index = (self._index + 1) % len(self._data)
		self._length = min(self._length + 1, len(self._data))
	
	
	def copy(self, dist: int, count: int, out: BinaryIO) -> None:
		"""Copies `count` bytes starting at `dist` bytes ago to the
		given output stream and also back into this buffer itself.
		Note that if the count exceeds the distance, then some of the output
		data will be a copy of data that was copied earlier in the process."""
		if count < 0 or not (1 <= dist <= self._length):
			raise ValueError("Invalid count or distance")
		
		readindex: int = (self._index - dist) % len(self._data)
		for _ in range(count):
			b: int = self._data[readindex]
			readindex = (readindex + 1) % len(self._data)
			out.write(bytes((b,)))
			self.append(b)



class Decompressor:
	
	# ---- Public functions ----
	
	"""Decompresses raw DEFLATE data (without zlib or gzip container) into bytes."""
	
	@staticmethod
	def decompress_to_bytes(bitin: BitInputStream) -> bytes:
		"""Reads from the given input stream, decompresses the data, and returns a new byte list."""
		out = io.BytesIO()
		Decompressor.decompress_to_stream(bitin, out)
		return out.getvalue()
	
	
	@staticmethod
	def decompress_to_stream(bitin: BitInputStream, out: BinaryIO) -> None:
		"""Reads from the given input stream, decompresses
		the data, and writes to the given output stream."""
		Decompressor(bitin, out)
	
	
	# ---- Private implementation ----
	
	# -- Fields --
	
	_input: BitInputStream
	_output: BinaryIO
	_dictionary: ByteHistory
	
	
	# -- Constructor --
	
	def __init__(self, bitin: BitInputStream, out: BinaryIO):
		"""Constructor, which immediately performs decompression"""
		
		# Initialize fields
		self._input = bitin
		self._output = out
		self._dictionary = ByteHistory(32 * 1024)
		
		# Process the stream of blocks
		while True:
			# Read the block header
			isfinal: bool = bitin.read_uint(1) != 0  # bfinal
			type: int = self._input.read_uint(2)  # btype
			
			# Decompress rest of block based on the type
			if type == 0:
				self._decompress_uncompressed_block()
			elif type == 1:
				self._decompress_huffman_block(Decompressor._FIXED_LITERAL_LENGTH_CODE, Decompressor._FIXED_DISTANCE_CODE)
			elif type == 2:
				litlencode, distcode = self._decode_huffman_codes()
				self._decompress_huffman_block(litlencode, distcode)
			elif type == 3:
				raise ValueError("Reserved block type")
			else:
				assert False, "Unreachable value"
			if isfinal:
				break
	
	
	# -- Constants: The code trees for static Huffman codes (btype = 1) --
	
	_FIXED_LITERAL_LENGTH_CODE = CanonicalCode([8]*144 + [9]*112 + [7]*24 + [8]*8)
	
	_FIXED_DISTANCE_CODE = CanonicalCode([5] * 32)
	
	
	# -- Method: Reading and decoding dynamic Huffman codes (btype = 2) --
	
	# Reads from the bit input stream, decodes the Huffman code
	# specifications into code trees, and returns the trees.
	def _decode_huffman_codes(self) -> Tuple[CanonicalCode,Optional[CanonicalCode]]:
		numlitlencodes: int = self._input.read_uint(5) + 257  # hlit + 257
		numdistcodes: int = self._input.read_uint(5) + 1      # hdist + 1
		
		numcodelencodes: int = self._input.read_uint(4) + 4   # hclen + 4
		codelencodelen: List[int] = [0] * 19  # This list is filled in a strange order
		codelencodelen[16] = self._input.read_uint(3)
		codelencodelen[17] = self._input.read_uint(3)
		codelencodelen[18] = self._input.read_uint(3)
		codelencodelen[ 0] = self._input.read_uint(3)
		for i in range(numcodelencodes - 4):
			j = (8 + i // 2) if (i % 2 == 0) else (7 - i // 2)
			codelencodelen[j] = self._input.read_uint(3)
		
		# Create the code length code
		codelencode = CanonicalCode(codelencodelen)
		
		# Read the main code lengths and handle runs
		codelens: List[int] = []
		while len(codelens) < numlitlencodes + numdistcodes:
			sym: int = codelencode.decode_next_symbol(self._input)
			if 0 <= sym <= 15:
				codelens.append(sym)
			elif sym == 16:
				if len(codelens) == 0:
					raise ValueError("No code length value to copy")
				runlen: int = self._input.read_uint(2) + 3
				codelens.extend(codelens[-1 : ] * runlen)
			elif sym == 17:
				runlen = self._input.read_uint(3) + 3
				codelens.extend([0] * runlen)
			elif sym == 18:
				runlen = self._input.read_uint(7) + 11
				codelens.extend([0] * runlen)
			else:
				assert False, "Symbol out of range"
		if len(codelens) > numlitlencodes + numdistcodes:
			raise ValueError("Run exceeds number of codes")
		
		# Create literal-length code tree
		litlencode = CanonicalCode(codelens[ : numlitlencodes])
		
		# Create distance code tree with some extra processing
		distcodelen: List[int] = codelens[numlitlencodes : ]
		if len(distcodelen) == 1 and distcodelen[0] == 0:
			distcode: Optional[CanonicalCode] = None  # Empty distance code; the block shall be all literal symbols
		else:
			# Get statistics for upcoming logic
			onecount: int = sum(1 for x in distcodelen if x == 1)
			otherpositivecount: int = sum(1 for x in distcodelen if x > 1)
			
			# Handle the case where only one distance code is defined
			if onecount == 1 and otherpositivecount == 0:
				# Add a dummy invalid code to make the Huffman tree complete
				distcodelen.extend([0] * (32 - len(distcodelen)))
				distcodelen[31] = 1
			distcode = CanonicalCode(distcodelen)
		
		return (litlencode, distcode)
	
	
	# -- Methods: Block decompression --
	
	# Handles and copies an uncompressed block from the bit input stream.
	def _decompress_uncompressed_block(self) -> None:
		# Discard bits to align to byte boundary
		while self._input.get_bit_position() != 0:
			self._input.read_uint(1)
		
		# Read length
		len : int = self._input.read_uint(16)
		nlen: int = self._input.read_uint(16)
		if len ^ 0xFFFF != nlen:
			raise ValueError("Invalid length in uncompressed block")
		
		# Copy bytes
		for _ in range(len):
			b: int = self._input.read_uint(8)  # Byte is aligned
			if b == -1:
				raise EOFError("Unexpected end of stream")
			self._output.write(bytes((b,)))
			self._dictionary.append(b)
	
	
	# Decompresses a Huffman-coded block from the bit input stream based on the given Huffman codes.
	def _decompress_huffman_block(self, litlencode: CanonicalCode, distcode: Optional[CanonicalCode]) -> None:
		# litlencode cannot be None, but distcode is allowed to be None
		while True:
			sym: int = litlencode.decode_next_symbol(self._input)
			if sym == 256:  # End of block
				break
			
			if sym < 256:  # Literal byte
				self._output.write(bytes((sym,)))
				self._dictionary.append(sym)
			else:  # Length and distance for copying
				run: int = self._decode_run_length(sym)
				assert 3 <= run <= 258, "Invalid run length"
				if distcode is None:
					raise ValueError("Length symbol encountered with empty distance code")
				distsym: int = distcode.decode_next_symbol(self._input)
				dist: int = self._decode_distance(distsym)
				assert 1 <= dist <= 32768, "Invalid distance"
				self._dictionary.copy(dist, run, self._output)
	
	
	# -- Methods: Symbol decoding --
	
	# Returns the run length based on the given symbol and possibly reading more bits.
	def _decode_run_length(self, sym: int) -> int:
		# Symbols outside the range cannot occur in the bit stream;
		# they would indicate that the decompressor is buggy
		assert 257 <= sym <= 287, f"Invalid run length symbol: {sym}"
		
		if sym <= 264:
			return sym - 254
		elif sym <= 284:
			numextrabits: int = (sym - 261) // 4
			return (((sym - 265) % 4 + 4) << numextrabits) + 3 + self._input.read_uint(numextrabits)
		elif sym == 285:
			return 258
		else:  # sym is 286 or 287
			raise ValueError(f"Reserved length symbol: {sym}")
	
	
	# Returns the distance based on the given symbol and possibly reading more bits.
	def _decode_distance(self, sym: int) -> int:
		# Symbols outside the range cannot occur in the bit stream;
		# they would indicate that the decompressor is buggy
		assert 0 <= sym <= 31, f"Invalid distance symbol: {sym}"
		
		if sym <= 3:
			return sym + 1
		elif sym <= 29:
			numextrabits: int = sym // 2 - 1
			return ((sym % 2 + 2) << numextrabits) + 1 + self._input.read_uint(numextrabits)
		else:  # sym is 30 or 31
			raise ValueError(f"Reserved distance symbol: {sym}")

File: typescript/deflate-decompress.ts
/* 
 * Simple DEFLATE decompressor (TypeScript)
 * 
 * Copyright (c) Project Nayuki
 * MIT License. See readme file.
 * https://www.nayuki.io/page/simple-deflate-decompressor
 */


namespace deflate {
	
	type byte = number;
	type int = number;
	
	
	/* 
	 * A canonical Huffman code, where the code values for each symbol is
	 * derived from a given sequence of code lengths. This data structure is
	 * immutable. This could be transformed into an explicit Huffman code tree.
	 * 
	 * Example:
	 *   Code lengths (canonical code):
	 *     Symbol A: 1
	 *     Symbol B: 0 (no code)
	 *     Symbol C: 3
	 *     Symbol D: 2
	 *     Symbol E: 3
	 *  
	 *   Generated Huffman codes:
	 *     Symbol A: 0
	 *     Symbol B: (Absent)
	 *     Symbol C: 110
	 *     Symbol D: 10
	 *     Symbol E: 111
	 *  
	 *   Huffman code tree:
	 *       .
	 *      / \
	 *     A   .
	 *        / \
	 *       D   .
	 *          / \
	 *         C   E
	 */
	class CanonicalCode {
		
		// This dictionary maps Huffman codes to symbol values. Each key is the
		// Huffman code padded with a 1 bit at the beginning to disambiguate codes
		// of different lengths (e.g. otherwise we can't distinguish 0b01 from
		// 0b0001). For the example of codeLengths=[1,0,3,2,3], we would have:
		//     0b1_0 -> 0
		//    0b1_10 -> 3
		//   0b1_110 -> 2
		//   0b1_111 -> 4
		private codeBitsToSymbol = new Map<int,int>();
		
		
		// Constructs a canonical Huffman code from the given list of symbol code lengths.
		// Each code length must be non-negative. Code length 0 means no code for the symbol.
		// The collection of code lengths must represent a proper full Huffman code tree.
		// Examples of code lengths that result in correct full Huffman code trees:
		// - [1, 1] (result: A=0, B=1)
		// - [2, 2, 1, 0, 0, 0] (result: A=10, B=11, C=0)
		// - [3, 3, 3, 3, 3, 3, 3, 3] (result: A=000, B=001, C=010, ..., H=111)
		// Examples of code lengths that result in under-full Huffman code trees:
		// - [0, 2, 0] (result: B=00, unused=01, unused=1)
		// - [0, 1, 0, 2] (result: B=0, D=10, unused=11)
		// Examples of code lengths that result in over-full Huffman code trees:
		// - [1, 1, 1] (result: A=0, B=1, C=overflow)
		// - [1, 1, 2, 2, 3, 3, 3, 3] (result: A=0, B=1, C=overflow, ...)
		public constructor(codeLengths: Readonly<Array<int>>) {
			let nextCode: int = 0;
			for (let codeLength = 1; codeLength <= CanonicalCode.MAX_CODE_LENGTH; codeLength++) {
				nextCode <<= 1;
				const startBit: int = 1 << codeLength;
				codeLengths.forEach((cl: int, symbol: int) => {
					if (cl != codeLength)
						return;
					if (nextCode >= startBit)
						throw new RangeError("This canonical code produces an over-full Huffman code tree");
					this.codeBitsToSymbol.set(startBit | nextCode, symbol);
					nextCode++;
				});
			}
			if (nextCode != 1 << CanonicalCode.MAX_CODE_LENGTH)
				throw new RangeError("This canonical code produces an under-full Huffman code tree");
		}
		
		
		// Decodes the next symbol from the given bit input stream based on this
		// canonical code. The returned symbol value is in the range [0, codeLengths.size()).
		public decodeNextSymbol(inp: BitInputStream): int {
			let codeBits: int = 1;
			while (true) {
				codeBits = codeBits << 1 | inp.readUint(1);
				const result: int|undefined = this.codeBitsToSymbol.get(codeBits);
				if (result !== undefined)
					return result;
			}
		}
		
		
		// The maximum Huffman code length allowed in the DEFLATE standard.
		private static readonly MAX_CODE_LENGTH: int = 15;
		
	}
	
	
	
	/* 
	 * Decompresses raw DEFLATE data (without zlib or gzip container) into bytes.
	 */
	export class Decompressor {
		
		// Reads from the given input stream, decompresses the data, and returns a new byte array.
		public static decompress(input: BitInputStream): Uint8Array {
			return new Uint8Array(new Decompressor(input).output);
		}
		
		
		private output: Array<byte> = [];
		
		private dictionary = new ByteHistory(32 * 1024);
		
		
		// Constructor, which immediately performs decompression.
		private constructor(
				private input: BitInputStream) {
			
			// Process the stream of blocks
			let isFinal: boolean;
			do {
				// Read the block header
				isFinal = this.input.readUint(1) != 0;  // bfinal
				const type: int = this.input.readUint(2);  // btype
				
				// Decompress rest of block based on the type
				if (type == 0)
					this.decompressUncompressedBlock();
				else if (type == 1)
					this.decompressHuffmanBlock(Decompressor.FIXED_LITERAL_LENGTH_CODE, Decompressor.FIXED_DISTANCE_CODE);
				else if (type == 2) {
					const [litLenCode, distCode]: [CanonicalCode,CanonicalCode|null] = this.decodeHuffmanCodes();
					this.decompressHuffmanBlock(litLenCode, distCode);
				} else if (type == 3)
					throw new Error("Reserved block type");
				else
					throw new Error("Unreachable value");
			} while (!isFinal);
			
			// Discard bits to align to byte boundary
			while (this.input.getBitPosition() != 0)
				this.input.readUint(1);
		}
		
		
		private static FIXED_LITERAL_LENGTH_CODE: CanonicalCode = Decompressor.makeFixedLiteralLengthCode();
		
		private static makeFixedLiteralLengthCode(): CanonicalCode {
			let codeLens: Array<int> = [];
			for (let i = 0; i < 144; i++) codeLens.push(8);
			for (let i = 0; i < 112; i++) codeLens.push(9);
			for (let i = 0; i <  24; i++) codeLens.push(7);
			for (let i = 0; i <   8; i++) codeLens.push(8);
			return new CanonicalCode(codeLens);
		}
		
		
		private static FIXED_DISTANCE_CODE: CanonicalCode = Decompressor.makeFixedDistanceCode();
		
		private static makeFixedDistanceCode(): CanonicalCode {
			let codeLens: Array<int> = [];
			for (let i = 0; i < 32; i++) codeLens.push(5);
			return new CanonicalCode(codeLens);
		}
		
		
		// Reads from the bit input stream, decodes the Huffman code
		// specifications into code trees, and returns the trees.
		private decodeHuffmanCodes(): [CanonicalCode,CanonicalCode|null] {
			const numLitLenCodes: int = this.input.readUint(5) + 257;  // hlit + 257
			const numDistCodes: int = this.input.readUint(5) + 1;      // hdist + 1
			
			// Read the code length code lengths
			const numCodeLenCodes: int = this.input.readUint(4) + 4;   // hclen + 4
			let codeLenCodeLen: Array<int> = [];  // This array is filled in a strange order
			for (let i = 0; i < 19; i++)
				codeLenCodeLen.push(0);
			codeLenCodeLen[16] = this.input.readUint(3);
			codeLenCodeLen[17] = this.input.readUint(3);
			codeLenCodeLen[18] = this.input.readUint(3);
			codeLenCodeLen[ 0] = this.input.readUint(3);
			for (let i = 0; i < numCodeLenCodes - 4; i++) {
				const j: int = (i % 2 == 0) ? (8 + Math.floor(i / 2)) : (7 - Math.floor(i / 2));
				codeLenCodeLen[j] = this.input.readUint(3);
			}
			
			// Create the code length code
			const codeLenCode = new CanonicalCode(codeLenCodeLen);
			
			// Read the main code lengths and handle runs
			let codeLens: Array<int> = [];
			while (codeLens.length < numLitLenCodes + numDistCodes) {
				const sym: int = codeLenCode.decodeNextSymbol(this.input);
				if (0 <= sym && sym <= 15)
					codeLens.push(sym);
				else if (sym == 16) {
					if (codeLens.length == 0)
						throw new Error("No code length value to copy");
					const runLen: int = this.input.readUint(2) + 3;
					for (let i = 0; i < runLen; i++)
						codeLens.push(codeLens[codeLens.length - 1]);
				} else if (sym == 17) {
					const runLen: int = this.input.readUint(3) + 3;
					for (let i = 0; i < runLen; i++)
						codeLens.push(0);
				} else if (sym == 18) {
					const runLen: int = this.input.readUint(7) + 11;
					for (let i = 0; i < runLen; i++)
						codeLens.push(0);
				} else
					throw new Error("Symbol out of range");
			}
			if (codeLens.length > numLitLenCodes + numDistCodes)
				throw new Error("Run exceeds number of codes");
			
			// Create literal-length code tree
			const litLenCodeLen: Array<int> = codeLens.slice(0, numLitLenCodes);
			const litLenCode = new CanonicalCode(litLenCodeLen);
			
			// Create distance code tree with some extra processing
			let distCodeLen: Array<int> = codeLens.slice(numLitLenCodes);
			let distCode: CanonicalCode|null;
			if (distCodeLen.length == 1 && distCodeLen[0] == 0)
				distCode = null;  // Empty distance code; the block shall be all literal symbols
			else {
				// Get statistics for upcoming logic
				const oneCount: int = distCodeLen.filter(x => x == 1).length;
				const otherPositiveCount: int = distCodeLen.filter(x => x > 1).length;
				
				// Handle the case where only one distance code is defined
				if (oneCount == 1 && otherPositiveCount == 0) {
					while (distCodeLen.length < 32)
						distCodeLen.push(0);
					distCodeLen[31] = 1;
				}
				distCode = new CanonicalCode(distCodeLen);
			}
			return [litLenCode, distCode];
		}
		
		
		// Handles and copies an uncompressed block from the bit input stream.
		private decompressUncompressedBlock(): void {
			// Discard bits to align to byte boundary
			while (this.input.getBitPosition() != 0)
				this.input.readUint(1);
			
			// Read length
			const  len: int = this.input.readUint(16);
			const nlen: int = this.input.readUint(16);
			if ((len ^ 0xFFFF) != nlen)
				throw new Error("Invalid length in uncompressed block");
			
			// Copy bytes
			for (let i = 0; i < len; i++) {
				const b: byte = this.input.readUint(8);  // Byte is aligned
				this.output.push(b);
				this.dictionary.append(b);
			}
		}
		
		
		// Decompresses a Huffman-coded block from the bit input stream based on the given Huffman codes.
		private decompressHuffmanBlock(litLenCode: CanonicalCode, distCode: CanonicalCode|null): void {
			while (true) {
				const sym: int = litLenCode.decodeNextSymbol(this.input);
				if (sym == 256)  // End of block
					break;
				
				else if (sym < 256) {  // Literal byte
					this.output.push(sym);
					this.dictionary.append(sym);
				} else {  // Length and distance for copying
					const run: int = this.decodeRunLength(sym);
					if (!(3 <= run && run <= 258))
						throw new Error("Invalid run length");
					if (distCode === null)
						throw new Error("Length symbol encountered with empty distance code");
					const distSym: int = distCode.decodeNextSymbol(this.input);
					const dist: int = this.decodeDistance(distSym);
					if (!(1 <= dist && dist <= 32768))
						throw new Error("Invalid distance");
					this.dictionary.copy(dist, run, this.output);
				}
			}
		}
		
		
		// Returns the run length based on the given symbol and possibly reading more bits.
		private decodeRunLength(sym: int): int {
			// Symbols outside the range cannot occur in the bit stream;
			// they would indicate that the decompressor is buggy
			if (!(257 <= sym && sym <= 287))
				throw new RangeError("Invalid run length symbol");
			
			if (sym <= 264)
				return sym - 254;
			else if (sym <= 284) {
				const numExtraBits: int = Math.floor((sym - 261) / 4);
				return (((sym - 265) % 4 + 4) << numExtraBits) + 3 + this.input.readUint(numExtraBits);
			} else if (sym == 285)
				return 258;
			else  // sym is 286 or 287
				throw new RangeError("Reserved length symbol");
		}
		
		
		// Returns the distance based on the given symbol and possibly reading more bits.
		private decodeDistance(sym: int): int {
			// Symbols outside the range cannot occur in the bit stream;
			// they would indicate that the decompressor is buggy
			if (!(0 <= sym && sym <= 31))
				throw new RangeError("Invalid distance symbol");
			
			if (sym <= 3)
				return sym + 1;
			else if (sym <= 29) {
				const numExtraBits: int = Math.floor(sym / 2) - 1;
				return ((sym % 2 + 2) << numExtraBits) + 1 + this.input.readUint(numExtraBits);
			} else  // sym is 30 or 31
				throw new RangeError("Reserved distance symbol");
		}
		
	}
	
	
	
	/* 
	 * Stores a finite recent history of a byte stream. Useful as an implicit
	 * dictionary for Lempel-Ziv schemes.
	 */
	class ByteHistory {
		
		// Maximum number of bytes stored in this history.
		private size: int;
		
		// Circular buffer of byte data.
		private data: Array<byte> = [];
		
		// Index of next byte to write to, always in the range [0, data.length).
		private index: int = 0;
		
		
		// Constructs a byte history of the given size.
		public constructor(size: int) {
			if (size < 1)
				throw new RangeError("Size must be positive");
			this.size = size;
		}
		
		
		// Appends the given byte to this history.
		// This overwrites the byte value at `size` positions ago.
		public append(b: byte): void {
			if (this.data.length < this.size)
				this.data.push(0);  // Dummy value
			if (!(0 <= this.index && this.index < this.data.length))
				throw new Error("Unreachable state");
			this.data[this.index] = b;
			this.index = (this.index + 1) % this.size;
		}
		
		
		// Copies `len` bytes starting at `dist` bytes ago to the
		// given output array and also back into this buffer itself.
		// Note that if the count exceeds the distance, then some of the output
		// data will be a copy of data that was copied earlier in the process.
		public copy(dist: int, len: int, out: Array<byte>): void {
			if (len < 0 || !(1 <= dist && dist <= this.data.length))
				throw new RangeError("Invalid length or distance");
			let readIndex: int = (this.index + this.size - dist) % this.size;
			for (let i = 0; i < len; i++) {
				const b: byte = this.data[readIndex];
				readIndex = (readIndex + 1) % this.size;
				out.push(b);
				this.append(b);
			}
		}
		
	}
	
	
	
	/* 
	 * A stream of bits that can be read. Because they come from an underlying byte stream, the
	 * total number of bits is always a multiple of 8. Bits are packed in little endian within
	 * a byte. For example, the byte 0x87 reads as the sequence of bits [1,1,1,0,0,0,0,1].
	 */
	export class BitInputStream {
		
		// In the range [0, data.length*8].
		private bitIndex: int = 0;
		
		
		// Constructs a bit input stream based on the given byte array.
		public constructor(
			private data: Uint8Array) {}
		
		
		// Returns the current bit position, which ascends from 0 to 7 as bits are read.
		public getBitPosition(): int {
			return this.bitIndex % 8;
		}
		
		
		// Reads a bit from this stream. Returns 0 or 1 if a bit is available, or -1 if
		// the end of stream is reached. The end of stream always occurs on a byte boundary.
		public readBitMaybe(): -1|0|1 {
			const byteIndex: int = this.bitIndex >>> 3;
			if (byteIndex >= this.data.length)
				return -1;
			const result = ((this.data[byteIndex] >>> (this.bitIndex & 7)) & 1) as (0|1);
			this.bitIndex++;
			return result;
		}
		
		
		// Reads the given number of bits from this stream,
		// packing them in little endian as an unsigned integer.
		public readUint(numBits: int): int {
			if (numBits < 0 || numBits > 31)
				throw new RangeError("Number of bits out of range");
			let result: int = 0;
			for (let i = 0; i < numBits; i++) {
				const bit: -1|0|1 = this.readBitMaybe();
				if (bit == -1)
					throw new Error("Unexpected end of data");
				result |= bit << i;
			}
			return result;
		}
		
	}
	
}
